modules/core/src/main/kotlin/com/metalbear/mirrord/MirrordBinaryManager.kt

```diff
@@ -0,0 +1,195 @@
+package com.metalbear.mirrord
+
+import com.intellij.openapi.progress.ProgressIndicator
+import com.intellij.openapi.progress.ProgressManager
+import com.intellij.openapi.progress.Task
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.util.SystemInfo
+import com.intellij.util.system.CpuArch
+import java.nio.charset.Charset
+import java.nio.file.Path
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.TimeUnit
+import java.net.URI
+import java.net.http.HttpClient
+import java.net.http.HttpRequest
+import java.net.http.HttpResponse
+import java.net.URLEncoder
+import java.nio.file.Files
+import java.time.Duration
+
+/**
+ * For dynamically fetching and storing mirrord binary.
+ */
+object MirrordBinaryManager {
+    private const val cliBinary = "mirrord"
+    private const val versionEndpoint = "https://version.mirrord.dev/v1/version"
+    private const val downloadEndpoint = "https://github.com/metalbear-co/mirrord/releases/download"
+
+    /**
+     * @return version of mirrord binary
+     *
+     * @throws RuntimeException if `--version` command failed
+     */
+    private fun getVersion(cliPath: String): String {
+        val child = Runtime.getRuntime().exec(arrayOf(cliPath, "--version"))
+
+        val result = child.waitFor()
+        if (result != 0) {
+            MirrordLogger.logger.debug("`mirrord --version` failed with code $result")
+            throw RuntimeException("failed to get mirrord version")
+        }
+
+        return child.inputReader().readLine().split(' ')[1].trim()
+    }
+
+    /**
+     * @return path to the local installation of mirrord
+     */
+    private fun getLocalBinary(requiredVersion: String?): String? {
+        try {
+            val child = Runtime.getRuntime().exec(arrayOf("which", "mirrord"))
```

Comment:
this wouldn't work with WSL - we probably need to pass the WSL stuff here then use it to execute and do this flow.

Reply:
Hmm, what do you mean? How should we handle WSL here?

Reply:
You're executing a process, if the IDE is running on Windows, we'd fail to do this.
The caller passes `wslDistribution` which can then be used to execute processes in the Linux context.
Maybe we should have our own wrapper so we pass it and it uses wslDistribution when needed..

Reply:
I made an update to this. No way of checking if it actually works on WSL though...